rust/hyperlane-base/src/types/multisig.rs

```diff
@@ -171,51 +203,218 @@ impl MultisigCheckpointSyncer {
+                    let signer =
+                        self.fetch_or_recover(&signed_checkpoint, &mut recovered_addresses)?;
 
-                    if H256::from(signer) != *validator {
+                    if signer != vw.validator {
                         debug!(
-                            validator = format!("{:#x}", validator),
+                            validator = format!("{:#x}", vw.validator),
                             index = index,
                             "Checkpoint signature mismatch"
                         );
                         continue;
                     }
 
+                    // let checkpoint_hash = signed_checkpoint.value.signing_hash();
+
                     // Push the signed checkpoint into the hashmap
                     let root = signed_checkpoint.value.root;
                     let signed_checkpoints = signed_checkpoints_per_root.entry(root).or_default();
                     signed_checkpoints.push(signed_checkpoint);
 
+                    let cumulative_weight: Weight = signed_checkpoints
+                        .iter()
+                        .filter_map(|sc| {
+                            let signer =
+                                self.fetch_or_recover(sc, &mut recovered_addresses).ok()?;
+                            weighted_validators
+                                .iter()
+                                .find(|vw| vw.validator == signer)
+                                .map(|vw| vw.weight)
+                        })
+                        .sum();
+
                     // Count the number of signatures for this signed checkpoint
                     let signature_count = signed_checkpoints.len();
                     debug!(
-                        validator = format!("{:#x}", validator),
+                        validator = format!("{:#x}", vw.validator),
                         index = index,
                         root = format!("{:#x}", root),
                         signature_count = signature_count,
+                        cumulative_weight = cumulative_weight,
                         "Found signed checkpoint"
                     );
 
-                    // If we've hit a quorum, create a MultisigSignedCheckpoint
-                    if signature_count >= threshold {
+                    // If we've hit a quorum in weight, create a MultisigSignedCheckpoint
+                    if cumulative_weight >= threshold_weight {
+                        // to conform to the onchain ordering of the set by address
+                        signed_checkpoints.sort_by_key(|sc| {
+                            self.fetch_or_recover(sc, &mut recovered_addresses)
+                                .ok()
+                                .and_then(|signer| {
+                                    weighted_validators
+                                        .iter()
+                                        .position(|vw| vw.validator == signer)
```

Comment:
ditto I don't think we should have O(n) operations / anything related to recovering in the sorting fn

I'd probably restructure this function a bit to be something like this instead of having `fetch_or_recover`:
1. Create
```
struct ValidatorSignedCheckpoint {
  signed_checkpoint: SignedCheckpointWithMessageId,
  validator: H256,
  weight: Weight,
  ism_index: usize,
}

#[derive(Default)]
struct ValidatorSignedCheckpoints {
  cumulative_weight: Weight,
  signed_checkpoints: Vec<ValidatorSignedCheckpoint>,
}

impl ValidatorSignedCheckpoints {
  fn push(...) { 
    // push into `self.signed_checkpoints`, update the cumulative weight
  }
}
```

2. Change `signed_checkpoints_per_root` to use this
```
let mut signed_checkpoints_per_root: HashMap<H256, ValidatorSignedCheckpoints> =
            HashMap::new();
```

3. Now `cumulative_weight` can come directly from the map entry in O(1)
4. Now the sorting to match the ISM order here can just use the `ism_index` for a basic O(nlog(n)):
```
validator_signed_checkpoints.signed_checkpoints.sort_by_key(|sc| sc.ism_index)
```